### DISCUSSION TITLE
zh/integrations/amazon-sagemaker/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9731] [Docs]
# zh/integrations/amazon-sagemaker/

探索Ultralytics YOLOv8 ，这是一种先进的实时对象检测和图像分割模型，适用于各种应用和硬件平台。

https://docs.ultralytics.com/zh/integrations/amazon-sagemaker/

**Comment by lijoe123** [comment]
I want to upload the YOLOv8 model I trained on my own dataset to Jupyter, and then compress it into a `model.tar.gz` file following the official `deployendpoint` steps. However, when I use `testendpoint`, I encounter an error. If I rename my trained model to `yolov8`, the test succeeds, but it uses the original model from the official website instead of my custom one. How should I modify `inference.py` to ensure it loads my trained model?

**Reply by glenn-jocher** [reply]
@lijoe123 to ensure your custom YOLOv8 model is used, modify the `inference.py` to load your model by specifying its path directly. Ensure the model name in your deployment matches your custom model's name. For detailed steps, refer to the [Ultralytics documentation on Amazon SageMaker](https://docs.ultralytics.com/integrations/amazon-sagemaker/).

**Reply by lijoe123** [reply]
Thank you for your answer. I had followed the `1_DeployEndpoint.ipynb` to delpoy my own model. 
In the **1.4 Create the SageMaker PyTorchModel**, I had set the `env={'TS_MAX_RESPONSE_SIZE':'20000000', 'MY_MODEL': 'my_own_model'}`. 
And in the `inference.py`:
```
def model_fn(model_dir):
    print("Executing model_fn from inference.py ...")
    env = os.environ
    model = YOLO(os.path.join(model_dir, env['MY_MODEL']))
    return model
```
I think it will be use my own model directly. But it failed, and it occur this problem:
```
ModelError: An error occurred (ModelError) when calling the InvokeEndpoint operation: Received server error (0)from primary with message "Your invocation timed out while waiting for a response from container primary. Reviewthe latency metrics for each container in Amazon Cloudatch, resolve the issue, and try again.". See https://useast-1. console. aws. aazon. com/cloudwatch/home?region=us-east-1#logEventViewer: group=/aws/sagemaker Endpoints/yolov8-pytorch-2024-09-06-11-33-49-418696 in account 211125332801 for more information.
```
Could you tell me how to modify the `inference.py` to load my own model. Thank you so much.

**Reply by glenn-jocher** [reply]
To ensure your custom YOLOv8 model is used, verify that the `model_fn` in `inference.py` correctly points to your model's path. Ensure the `MY_MODEL` environment variable matches your model's filename. If issues persist, check CloudWatch logs for detailed error messages. For further guidance, refer to the [Ultralytics documentation on Amazon SageMaker](https://docs.ultralytics.com/integrations/amazon-sagemaker/).

**Comment by SJTUPanda** [comment]
Thanks for your sharing, it's very clear and useful. My own deployment workflow on AWS has already built but the test result is not so satisfying. The performance of detection result of yolo11-xl on AWS is worse than on my own laptop. I not sure if there is some default  quantization when it's deployed to the AWS(I did not modified much to host-yolov8-on-sagemaker-endpoint), or do you have some other ideas about which would bring this significant performance differences？

**Reply by glenn-jocher** [reply]
@SJTUPanda thank you for your feedback! The performance difference could stem from several factors like hardware disparities (AWS instance vs laptop GPU), inference settings, or environment configurations. First, verify that your SageMaker instance type matches your local hardware capabilities, and ensure identical inference parameters (e.g., `imgsz`, `batch`). Check our [Amazon SageMaker deployment guide](https://docs.ultralytics.com/integrations/amazon-sagemaker/) for optimization tips and consider profiling both environments with `yolo val` to isolate bottlenecks. Let us know if specific metrics differ! 🚀

**Reply by SJTUPanda** [reply]
Thanks for your quick reply. The issue is fixed after I comment out resize related code. Although the model is trained with (640,640) and the input image size is (720,440), using the raw images directly as input is better than resized it. I am curious the input images would be auto-resized by the model just as in the training/testing part? And why resized by myself has so much impact on the results? If these part code should be update because yolo version updating?
```python
image_height, image_width, _ = orig_image.shape
model_height, model_width = 640, 640
x_ratio = image_width/model_width
y_ratio = image_height/model_height

# Resize the image as numpy array
resized_image = cv2.resize(orig_image, (model_height, model_width))
# Conver the array into jpeg
resized_jpeg = cv2.imencode('.jpg', resized_image)[1]
# Serialize the jpg using base 64
payload = base64.b64encode(resized_jpeg).decode('utf-8')

**Reply by glenn-jocher** [reply]
Thank you for sharing your findings! YOLO models do automatically resize inputs to match their training dimensions while preserving aspect ratio with letterboxing. Manual resizing can distort aspect ratios or introduce artifacts, potentially degrading performance. We recommend using the model's native preprocessing by feeding full-size images directly, as this maintains proper scaling and aspect ratio handling. For implementation details, refer to our [model deployment best practices](https://docs.ultralytics.com/guides/model-deployment-practices/) regarding input consistency.
